src/main/java/de/tum/in/www1/artemis/config/migration/entries/MigrationEntry20230808_203400.java

```diff
@@ -0,0 +1,460 @@
+                    studentParticipation.setBuildPlanId(null);
+                    programmingExerciseStudentParticipationRepository.save(studentParticipation);
+                    return null;
+                }
+                return studentParticipation;
+            }).filter(Objects::nonNull).forEach(studentParticipation -> {
+                var startMs = System.currentTimeMillis();
+                List<AuxiliaryRepository> auxiliaryRepositories = new ArrayList<>();
+                if (ciMigrationService.orElseThrow().supportsAuxiliaryRepositories()) {
+                    auxiliaryRepositories = auxiliaryRepositoryRepository.findByExerciseId(studentParticipation.getProgrammingExercise().getId());
+                }
+                migrateStudentBuildPlan(studentParticipation.getProgrammingExercise(), studentParticipation, auxiliaryRepositories, errorMap);
+                log.info("Migrated student build plan for exercise {} in {}ms", studentParticipation.getProgrammingExercise().getId(), System.currentTimeMillis() - startMs);
+            });
+
+            log.info("Migrated {} / {} programming exercises in current thread", (Math.min(exerciseCount, batch + 1)), exerciseCount);
+            log.info("Migrated batch in {}ms", System.currentTimeMillis() - batchStart);
+        }
+    }
+
+    /**
+     * Migrates a single test build plan.
+     *
+     * @param exercise The exercise to migrate
+     * @param errorMap The error map to write errors to
+     */
+    private void migrateTestBuildPlan(ProgrammingExercise exercise, Map<ProgrammingExercise, Boolean> errorMap) {
+        if (exercise.getTestRepositoryUrl() == null) {
+            /*
+             * when the test repository url is null, we don't have to migrate the test build plan, saving multiple API calls
+             * this is also only needed for Jenkins, as Bamboo does not have a separate test build plan
+             */
+            return;
+        }
+        try {
+            ciMigrationService.orElseThrow().deleteBuildTriggers(null, null, exercise.getVcsTestRepositoryUrl());
+        }
+        catch (VersionControlException | ContinuousIntegrationException | NoSuchElementException | RestClientException e) {
+            log.warn("Failed to delete test build triggers for exercise {} with test repository {}", exercise.getId(), exercise.getVcsTestRepositoryUrl(), e);
+            errorMap.put(exercise, true);
+        }
+    }
+
+    /**
+     * Migrates a single solution build plan.
+     *
+     * @param exercise The exercise to migrate
+     * @param errorMap The error map to write errors to
+     */
+    private void migrateStudentBuildPlan(ProgrammingExercise exercise, ProgrammingExerciseStudentParticipation studentParticipation,
+            List<AuxiliaryRepository> auxiliaryRepositories, Map<ProgrammingExercise, Boolean> errorMap) {
+        VcsRepositoryUrl repositoryUrl;
+        try {
```

Comment:
I refer to this methods. I'm not sure whether we have to refactor the methods, but I think they are hard to read (I'm unsure how we could refactor the methods :) )

I'm not sure, but we could try a helper method like:

```java
private void migrationStep(ProgrammingEx .. , .. ,  .. errorMap, String nameOfMigrationStep, RunnableWithException migration) {
try {
migration();
}
catch(Exception e){
      log.warn(...., nameOfMigration);
      errorMap.put(....);
 } 
}
```

Then we could write something like:

```java
migrationStep(....., "SolutionPlan", ()->ciMigrationService.orElseThrow().overrideBuildPlanNotification(exercise.getProjectKey(), studentParticipation.getBuildPlanId(), repositoryUrl) );

```

Reply:
I agree, it's a bit hard to read with all the try-catch :)

Reply:
The callback solution won't work with the one case where control flow is changed with `return`, though 🤔 
```java
        catch (URISyntaxException e) {
            log.warn("Failed to convert git url {} for studentParticipationId {} exerciseId {} with buildPlanId {}, will abort migration for this Participation",
                    studentParticipation.getVcsRepositoryUrl(), studentParticipation.getId(), exercise.getId(), studentParticipation.getBuildPlanId(), e);
            errorMap.put(exercise, false);
            return;
        }
```

Reply:
We specifically implemented it like this so that we have each step that can fail on its own, and separate methods for template/solution/student for easier traceability in case there are errors.

Reply:
> The callback solution won't work with the one case where control flow is changed with `return`, though 🤔
> 
> ```java
>         catch (URISyntaxException e) {
>             log.warn("Failed to convert git url {} for studentParticipationId {} exerciseId {} with buildPlanId {}, will abort migration for this Participation",
>                     studentParticipation.getVcsRepositoryUrl(), studentParticipation.getId(), exercise.getId(), studentParticipation.getBuildPlanId(), e);
>             errorMap.put(exercise, false);
>             return;
>         }
> ```

Yes, but that's the initialisation of repositoryUrl as far as I can see .. I'd not count that as migration step :)

Reply:
> We specifically implemented it like this so that we have each step that can fail on its own, and separate methods for template/solution/student for easier traceability in case there are errors.

That would also be the case if you use a helper method for the try catch stuff :) (then you have to look into the second entry in the stack trace :))

Reply:
We don't need to focus on perfect code quality in the migration. Thank you for the suggestion, though.
As long as it does the right thing, it's fine. It will never change once merged and removed at some point.